opendevin/runtime/utils/runtime_build.py

```diff
@@ -47,57 +58,154 @@ def _put_source_code_to_dir(temp_dir: str) -> str:
+    shutil.copy(pyproject_path, os.path.join(temp_dir, 'pyproject.toml'))
+    logger.info(f'pyproject.toml copied to {temp_dir}')
+
     return filename
 
 
+def _generate_dockerfile_nodejs(
+    base_image: str, source_code_dirname: str, skip_init: bool = False
+) -> str:
+    """Generate the Dockerfile content for a python-nodejs image for the eventstream runtime."""
+    dockerfile_content = f'FROM {base_image}\n'
+    if not skip_init:
+        # Set important ENV vars from the get go
+        # Otherwise poetry will not install the environment correctly
+        dockerfile_content += (
+            'ENV DEBIAN_FRONTEND=noninteractive '
+            'POETRY_NO_INTERACTION=1 '
+            'POETRY_VIRTUALENVS_IN_PROJECT=1 '
+            'POETRY_VIRTUALENVS_CREATE=1 '
+            'POETRY_CACHE_DIR=/tmp/poetry_cache\n'
+        )
+        # not sure if we need to set this in the sandbox? --tobitege
+        # ENV PYTHONPATH='/opendevin/code'
+
+        dockerfile_content += (
+            # Create a policy-rc.d script to prevent services from starting
+            "RUN echo '#!/bin/sh\\nexit 101' > /usr/sbin/policy-rc.d && chmod +x /usr/sbin/policy-rc.d \n"
+            # Install necessary packages and clean up
+            'RUN apt-get update && apt-get install -y apt-utils\n'
+            'RUN apt-get update && apt-get install -y sudo curl vim\\\n'
+            ' nano zip build-essential jq iproute2 &&\\\n'
+            ' apt-get clean && rm -rf /var/lib/apt/lists/*\n'
+            # symlink python3 to python
+            'RUN ln -s /usr/bin/python3 /usr/bin/python\n'
+            # Create necessary directories
+            'RUN mkdir -p /opendevin && mkdir -p /opendevin/logs && chmod 777 /opendevin/logs\n'
+        )
+    # Copy the project directory to the container
+    dockerfile_content += 'COPY project.tar.gz /opendevin\n'
+
+    # Copy pyproject.toml to the source code directory
+    dockerfile_content += 'COPY pyproject.toml /opendevin/code/\n'
+
+    # Remove /opendevin/code if it exists
+    dockerfile_content += (
+        'RUN if [ -d /opendevin/code ]; then rm -rf /opendevin/code; fi\n'
+    )
+    # Unzip the tarball to /opendevin/code
+    dockerfile_content += (
+        'RUN cd /opendevin && tar -xzvf project.tar.gz && rm project.tar.gz\n'
+    )
+
+    # Install agentskills and browser dependencies (update if needed)
+    dockerfile_content += (
+        f'RUN mv /opendevin/{source_code_dirname} /opendevin/code && cd /opendevin/code && \\\n'
+        '    poetry install --only main && \\\n'  # run main poetry install to install all dependencies!
+        '    poetry run pip install \\\n'
+        '    jupyterlab notebook jupyter_kernel_gateway flake8 \\\n'
+        '    python-docx PyPDF2 python-pptx pylatexenc openai \\\n'
+        '    python-dotenv toml termcolor pydantic python-docx pyyaml \\\n'
+        '    docker pexpect tenacity e2b \\\n'
+        '    browsergym minio playwright \n'
```

Comment:
I actually think we don't need to install these manually because we already have these in the `pyproject.toml` - and it will automatically installed into the right version (`poetry.lock`), right? I guess this is why i like having poetry inside the runtime 😄

Reply:
Right, I had copied several over from the plugins code (jupyterlab, browsergym, playwright etc.).
I haven't checked if they're all in the pyproject file, but if so, they'd be "duplicates" without consequence.

Reply:
Oh, I don't think all of the browser stuff is in the pyproject, on 2nd thought.

Reply:
Yep, i think this is the only line that needed to be added:
https://github.com/OpenDevin/OpenDevin/blob/a61ac5a21425c21c0d76b213f03d18c80fbb5fad/Makefile#L152-L153